### DISCUSSION TITLE
How modular and flexible is the Toga for adding other backends (GUI)?

### DISCUSSION THREAD

**Discussion by Heus-Sueh** [beeware/toga#2253] [General]
Would it be too difficult to add other backends or is the problem the time it would take to accomplish this task?

**Comment by freakboy3742** [comment]
Toga has been specifically built so that backends are completely pluggable, and can be contributed or maintained by a third party.

To define a new backend, create a normal Python package; in that package's metadata, declare a dependency on `toga-core`, and register a `toga.backend` entry point. Using PEP621 metadata, you could expose a custom backend for Windows users with:
```
[project.entry-points."toga.backends"]
windows = "togax_custom_backend"
```

`togax_custom_backend` is the name of the python module that is provided by your Python package. **You should not use the `toga` prefix in your package name** - we're reserving that for *official* packages. User contribute packages should use the `togax` prefix.

This package should have a `togax_custom_backend.factory` module; that provides the namespace that Toga will use to construct platform-specific objects like `App`, `Window` and `Button`. 

As long as `toga-core` and `togax_custom_backend` are both installed on Windows, your custom backend will be used. If there is more than one valid backend for the current platform (e.g., if you had your custom backend *and* toga-winforms installed), you will need to specify which one to use, by setting ``TOGA_BACKEND=togax_custom_backend`` in your environment.
